### PR TITLE
[Graphbolt][CUDA] Migrate utils and cuda common from dgl

### PR DESCRIPTION
**Pull request by mfbalin**
In this PR:

1. Migrate some common CUDA-related kernel launch macros from DGL
2. Modify the existing CUDA code to utilize these CUDA kernel launch macros.
4. The CUDA stream is received from the torch runtime instead of utilizing the default CUDA stream.

### CODE REVIEW

File: graphbolt/src/cuda/common.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+inline bool is_zero(T size) {
+  return size == 0;
+}
+
+template <>
+inline bool is_zero<dim3>(dim3 size) {
+  return size.x == 0 || size.y == 0 || size.z == 0;
+}
+
+#define CUDA_DRIVER_CALL(x)                                             \
```
Comment (frozenbugs) [inline graphbolt/src/cuda/common.h:24]: This is not used in this PR, maybe add it when we need it?

File: graphbolt/src/cuda/utils.h [hunk 2]
```diff
@@ -31,6 +31,22 @@ inline int FindNumThreads(int size) {
   return ret;
 }
 
+template <typename T>
+int _NumberOfBits(const T& range) {
```
Comment (frozenbugs) [inline graphbolt/src/cuda/utils.h:35]: ditto.

File: graphbolt/src/cuda/common.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  {                                                                     \
+    CUresult result = x;                                                \
+    if (result != CUDA_SUCCESS && result != CUDA_ERROR_DEINITIALIZED) { \
+      const char* msg;                                                  \
+      cuGetErrorName(result, &msg);                                     \
+      LOG(FATAL) << "CUDAError: " #x " failed with error: " << msg;     \
+    }                                                                   \
+  }
+
+#define CUDA_CALL(func)                                      \
```
Comment (frozenbugs) [inline graphbolt/src/cuda/common.h:34]: ditto.

**Comment by frozenbugs** [comment]
LGTM, @czkkkkkk ptal asap.

**Review by frozenbugs — APPROVED** [review]

File: graphbolt/src/cuda/common.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  return size.x == 0 || size.y == 0 || size.z == 0;
+}
+
+#define CUDA_KERNEL_CALL(kernel, nblks, nthrs, shmem, stream, ...)    \
+  {                                                                   \
+    if (!graphbolt::cuda::is_zero((nblks)) &&                         \
+        !graphbolt::cuda::is_zero((nthrs))) {                         \
+      (kernel)<<<(nblks), (nthrs), (shmem), (stream)>>>(__VA_ARGS__); \
+      cudaError_t e = cudaGetLastError();                             \
+      CHECK(e == cudaSuccess || e == cudaErrorCudartUnloading)        \
```
[the remark this answers is not in the document]
Comment (mfbalin) [reply graphbolt/src/cuda/common.h:30]: When writing code, it is better to use the macros that we use in DGL, as then there won't be a chance that we forget to write the check. The `C10_CUDA_KERNEL_LAUNCH_CHECK` requires writing on a separate line and can be easy to forget going forward.

File: graphbolt/src/cuda/common.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/**
+ *  Copyright (c) 2017-2023 by Contributors
+ * @file cuda_common.h
```
[the remark this answers is not in the document]
Comment (mfbalin) [reply graphbolt/src/cuda/common.h]: Replaced with `cuda/common.h`. Do you think the filename is good as `cuda/common.h` or should it be `cuda/cuda_common.h`?

File: graphbolt/src/cuda/common.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (mfbalin) [reply graphbolt/src/cuda/common.h:30]: Oh, you mean the `CUDA_KERNEL_CALL` definition. I checked what `C10_CUDA_KERNEL_LAUNCH_CHECK` does and it looks almost identical.

[the remark this answers is not in the document]
Comment (mfbalin) [reply graphbolt/src/cuda/common.h:30]: Except that it doesn't check for whether there are nonzero blocks or threads.

File: graphbolt/src/cuda/common.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (czkkkkkk) [reply graphbolt/src/cuda/common.h]: Yeah. `cuda/common.h` is good.

File: graphbolt/src/cuda/common.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (czkkkkkk) [reply graphbolt/src/cuda/common.h:30]: I see. Then we can replaced L29-30 by `C10_CUDA_KERNEL_LAUNCH_CHECK` since it prints the `__FILE__` information.

**Review by czkkkkkk — APPROVED** [review]

**Pull request merged**